spec/unit/pops/types/p_object_type_spec.rb

```diff
@@ -0,0 +1,599 @@
+require 'spec_helper'
+require 'puppet/pops'
+
+module Puppet::Pops
+module Types
+describe 'The Object Type' do
+  let(:parser) { TypeParser.new }
+  let(:pp_parser) { Puppet::Pops::Parser::EvaluatingParser.new }
+  let(:loader) { Puppet::Pops::Loader::BaseLoader.new(nil, 'type_parser_unit_test_loader') }
+
+  def type_object_t(name, body_string)
+    TypeFactory.type_alias(name, pp_parser.parse_string("Object[{#{body_string}}]").current)
+  end
+
+  def expect_object(name, body_string)
+    obj = type_object_t(name, body_string)
+    loader.expects(:load).with(:type, name.downcase).at_least_once.returns obj
+    obj
+  end
+
+  def parse_object(name, body_string)
+    expect_object(name, body_string)
+    parser.parse(name, loader)
+  end
+
+  context 'when dealing with attributes' do
+    it 'raises an error when the attribute type is not a type' do
+      obj = <<-OBJECT
+        attributes => {
+          a => 23
```

Comment:
For constants it could work, but if we do that, then I'm not so the kind 'constant' should be implied because inferring could work for other kinds as well (the inferred type would need to be generalized).

Then again, yet another special rule to remember. So I'm ambivalent.